dex/src/main/java/io/crate/data/CompositeBatchIterator.java

```diff
@@ -0,0 +1,176 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.data;
+
+import io.crate.concurrent.CompletableFutures;
+
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.concurrent.CompletionStage;
+
+/**
+ * BatchIterator implementation that is backed by multiple other BatchIterators.
+ *
+ * It will prefer loaded BatchIterator's over unloaded iterators.
+ * It will consume each iterator to it's end before consuming the next iterator in order to make sure that repeated
+ * consumption via {@link #moveToStart()} always returns the same result.
```

Comment:
well probably in the future we could have different implementations if no scroll is required